src/rabbitmq.js

```diff
@@ -0,0 +1,360 @@
+   *
+   * @return {Promise} Promise resolved when all queues consuming.
+   */
+  consume (): Promise {
+    const log = this.log.child({ method: 'consume' })
+    log.info('starting to consume')
+    if (!this._isConnected()) {
+      return Promise.reject(new Error('you must .connect() before consuming'))
+    }
+    const subscriptions = this.subscriptions
+    this.subscriptions = new Immutable.Map()
+    const channel = this.channel
+    return Promise.map(subscriptions.keySeq(), (queue) => {
+      const handler = subscriptions.get(queue)
+      log.info({ queue: queue }, 'consuming on queue')
+      // XXX(bryan): is this valid? should I not be checking _this_.consuming?
+      if (this.consuming.has(queue)) {
+        log.warn({ queue: queue }, 'already consuming queue')
+        return true
+      }
+      function wrapper (msg) {
+        const job = JSON.parse(msg.content)
+        handler(job, () => {
+          channel.ack(msg)
+        })
+      }
+      return Promise.resolve(this.channel.consume(queue, wrapper))
+        .then((consumeInfo) => {
+          this.consuming = this.consuming.set(queue, consumeInfo.consumerTag)
+        })
+    })
+  }
+
+  /**
+   * Unsubscribe and stop consuming from all queues.
+   *
+   * @return {Promise} Promise resolved when all queues canceled.
+   */
+  unsubscribe (): Promise {
+    const consuming = this.consuming
+    return Promise.map(consuming.keySeq(), (queue) => {
+      const consumerTag = consuming.get(queue)
+      return Promise.resolve(this.channel.cancel(consumerTag))
+        .then(() => {
+          this.consuming = this.consuming.delete(queue)
+        })
+    })
+  }
+
+  /**
+   * Disconnect from RabbitMQ.
+   *
+   * @return {Promise} Promise resolved when disconnected from RabbitMQ.
+   */
+  disconnect (): Promise {
+    if (!this._isPartlyConnected()) {
+      return Promise.reject(new Error('not connected. cannot disconnect.'))
+    }
+    return Promise.resolve(this.connection.close())
+  }
+
+  // Private Methods
+
+  /**
+   * Error handler for the RabbitMQ connection.
+   *
+   * @private
+   * @throws Error
+   * @param {object} err Error object from event.
+   */
+  _connectionErrorHandler (err: Error) {
+    this.log.error({ err: err }, 'connection has caused an error')
+    throw err
+  }
+
+  /**
+   * Error handler for the RabbitMQ channel.
+   * @private
+   * @throws Error
+   * @param {object} err Error object from event.
+   */
+  _channelErrorHandler (err: Error) {
+    this.log.error({ err: err }, 'channel has caused an error')
+    throw err
+  }
+
+  /**
+   * Check to see if model is connected.
+   *
+   * @return {Boolean} True if model is connected and channel is established.
+   */
+  _isConnected (): boolean {
+    return !!(this._isPartlyConnected() && this.channel)
+  }
+
+  /**
+   * Check to see if model is _partially_ connected. This means that the
+   * connection was established, but the channel was not.
```

Comment:
connection was established, but the channel is not. I consider both a requirement to be 'connected'